{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "444ab4cd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-12T09:07:30.130961Z",
     "iopub.status.busy": "2024-06-12T09:07:30.130577Z",
     "iopub.status.idle": "2024-06-12T09:07:30.136124Z",
     "shell.execute_reply": "2024-06-12T09:07:30.135666Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "sys.path.append('/home/techt/Desktop/a4s')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca808c31",
   "metadata": {},
   "source": [
    "### Step 1: Initial Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e5ce571",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-12T09:07:30.138486Z",
     "iopub.status.busy": "2024-06-12T09:07:30.138325Z",
     "iopub.status.idle": "2024-06-12T09:07:30.464591Z",
     "shell.execute_reply": "2024-06-12T09:07:30.464213Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Background Information:\n",
      "!Series_title\t\"STAT3 phosphorylation at serine 727 activates specific genetic programs and promotes clear cell renal cell carcinoma (ccRCC) aggressiveness\"\n",
      "!Series_summary\t\"The signal transducer and activator of transcription 3 (STAT3) is a transcription factor mainly activated by phosphorylation in either tyrosine 705 (Y705) or serine 727 (S727) residues that regulates essential processes such as cell differentiation, apoptosis inhibition, or cell survival.\"\n",
      "!Series_summary\t\"we used microarrays to evaluate the effects of the STAT3 phosphomutants on global gene expression and identify the genes and pathways regulated by different STAT3 phosphorylation states in the 769-P cell line.\"\n",
      "!Series_overall_design\t\"we have generated human-derived ccRCC cell lines carrying STAT3 Y705 and S727 phosphomutants to identify genes and pathways regulated by pS727 that could be distinguished from those regulated by pY705 or by the combination of both. First, 769-P cells were depleted of endogenous STAT3 using shRNA and STAT3 WT form was then rescued. On this rescued STAT3 gene backbone, Y705 and S727 STAT3 phosphomutants were generated by introducing structurally similar amino acids that prevent (phosphoablative) or mimic (phosphomimetic) phosphorylation for each residue. A phosphomimetic substitution for Y705, however, was not possible since tyrosine is an aromatic amino acid and neither aspartic nor glutamic acid resembles the structure or charge density of a phosphotyrosine. To overcome this, we used interleukin-6 (IL6), a classic activator of the JAK/STAT3 pathway via pY705.\"\n",
      "Sample Characteristics Dictionary:\n",
      "{0: ['phenotype: Normal expression of endogenous STAT3', 'phenotype: Absence (reduction) of endogenous STAT3', 'phenotype: Normal activity of STAT3', 'phenotype: Tyr705 cannot be phosphorylated, free Ser727', 'phenotype: Ser727 cannot be phosphorylated, free Tyr705', 'phenotype: Ser727 artificially phosphorylated, free Tyr705', 'phenotype: Both, Tyr705 and Ser727, cannot be phosphorylated', 'phenotype: Tyr705 cannot be phosphorylated, Ser727 artificially phosphorylated']}\n"
     ]
    }
   ],
   "source": [
    "from utils.preprocess import *\n",
    "# 1. Identify the paths to the soft file and the matrix file\n",
    "cohort_dir = '/media/techt/DATA/GEO/Kidney_Clear_Cell_Carcinoma/GSE245862'\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "\n",
    "# 2. Read the matrix file to obtain background information and sample characteristics data\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "\n",
    "# 3. Obtain the sample characteristics dictionary from the clinical dataframe\n",
    "sample_characteristics_dict = get_unique_values_by_row(clinical_data)\n",
    "\n",
    "# 4. Explicitly print out all the background information and the sample characteristics dictionary\n",
    "print(\"Background Information:\")\n",
    "print(background_info)\n",
    "print(\"Sample Characteristics Dictionary:\")\n",
    "print(sample_characteristics_dict)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da142f8",
   "metadata": {},
   "source": [
    "### Step 2: Dataset Analysis and Clinical Feature Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1d9c141f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-12T09:07:30.466939Z",
     "iopub.status.busy": "2024-06-12T09:07:30.466795Z",
     "iopub.status.idle": "2024-06-12T09:07:30.471119Z",
     "shell.execute_reply": "2024-06-12T09:07:30.470820Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: ./preprocessed/Kidney_Clear_Cell_Carcinoma/cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "is_gene_available = False\n",
    "trait_row = age_row = gender_row = None  # set to different values when applicable\n",
    "convert_trait = convert_age = convert_age = None  # define the functions when applicable\n",
    "\n",
    "# Step 1: Determine if gene expression data is available\n",
    "is_gene_available = True\n",
    "\n",
    "# Step 2.1: Data Availability\n",
    "trait_row = None  # Assuming phenotype data doesn't directly capture Kidney_Clear_Cell_Carcinoma condition\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "# Step 2.3: Data Type Conversion Functions\n",
    "def convert_trait(value):\n",
    "    return None\n",
    "\n",
    "def convert_age(value):\n",
    "    try:\n",
    "        return float(value.split(\":\")[1].strip())\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "def convert_gender(value):\n",
    "    gender_map = {\"female\": 0, \"male\": 1}\n",
    "    try:\n",
    "        gender = value.split(\":\")[1].strip().lower()\n",
    "        return gender_map.get(gender, None)\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "# Step 3: Save Metadata Information\n",
    "save_cohort_info('GSE245862', './preprocessed/Kidney_Clear_Cell_Carcinoma/cohort_info.json', is_gene_available, trait_row is not None)\n",
    "\n",
    "# Since trait_row is None, we skip Step 4 as stated in the instructions.\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
