{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6fbf430c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:43:55.023769Z",
     "iopub.status.busy": "2025-03-25T04:43:55.023589Z",
     "iopub.status.idle": "2025-03-25T04:43:55.190716Z",
     "shell.execute_reply": "2025-03-25T04:43:55.190266Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '../..')))\n",
    "\n",
    "# Path Configuration\n",
    "from tools.preprocess import *\n",
    "\n",
    "# Processing context\n",
    "trait = \"Vitamin_D_Levels\"\n",
    "cohort = \"GSE34450\"\n",
    "\n",
    "# Input paths\n",
    "in_trait_dir = \"../../input/GEO/Vitamin_D_Levels\"\n",
    "in_cohort_dir = \"../../input/GEO/Vitamin_D_Levels/GSE34450\"\n",
    "\n",
    "# Output paths\n",
    "out_data_file = \"../../output/preprocess/Vitamin_D_Levels/GSE34450.csv\"\n",
    "out_gene_data_file = \"../../output/preprocess/Vitamin_D_Levels/gene_data/GSE34450.csv\"\n",
    "out_clinical_data_file = \"../../output/preprocess/Vitamin_D_Levels/clinical_data/GSE34450.csv\"\n",
    "json_path = \"../../output/preprocess/Vitamin_D_Levels/cohort_info.json\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "108593f8",
   "metadata": {},
   "source": [
    "### Step 1: Initial Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "941be9d7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:43:55.192453Z",
     "iopub.status.busy": "2025-03-25T04:43:55.192276Z",
     "iopub.status.idle": "2025-03-25T04:43:55.424921Z",
     "shell.execute_reply": "2025-03-25T04:43:55.424439Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files in the cohort directory:\n",
      "['GSE34450_family.soft.gz', 'GSE34450_series_matrix.txt.gz']\n",
      "Identified SOFT files: ['GSE34450_family.soft.gz']\n",
      "Identified matrix files: ['GSE34450_series_matrix.txt.gz']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Background Information:\n",
      "!Series_title\t\"Genes Associated with MUC5AC Expression in the Human Airway Epithelium\"\n",
      "!Series_summary\t\"To help define the genes associated with mucus synthesis and secretion in the human small airway epithelium, we hypothesized that comparison of the transcriptomes of the small airway epithelium of individuals that express high vs low levels of MUC5AC, a major secretory mucin and the major component of airway mucus, could be used as a probe to identify the genes related to human small airway mucus production / secretion.   Genome-wide comparison between healthy nonsmokers grouped as “high MUC5AC expressors” vs “low MUC5AC expressors” identified significantly up-regulated and  down-regulated genes in the high vs low expressors. Based on the literature, genes in the up-regulated list were used to identify a 73 “MUC5AC-associated core gene” list with 9 categories: mucus components; mucus-producing cell differentiation-related transcription factor; mucus-producing cell differentiation-related pathway or mediator; post-translational modification of mucin; vesicle transport; endoplasmic reticulum stress-related; secretory granule-associated; mucus secretion-related regulator and mucus hypersecretory-related ion channel.  The identification of the genes associated with increased small airway mucin production in humans should be useful in identifying therapeutic targets to treat small airway mucus hypersecretion.\"\n",
      "!Series_overall_design\t\"Since airway surface epithelial cells are the sole local source of mucus in the small airways, which are devoid of submucosal glands and mucus contributes to the morbidity and mortality of  smoking-related lung diseases, especially chronic obstructive pulmonary disease (COPD), we sought to define the genes associated with mucus synthesis and secretion in the human small airway epithelium.  Microarray analysis revealed 528  up-regulated and 15 down-regulated genes in MUC5AC high expressors compared to the MUC5AC low expressors. From 528 up-regulated genes, we found 73 genes with literature supported roles or potential roles in mucus production / secretion. These MUC5AC-core genes suggest that multiple molecular events involving the nucleus, endoplasmic reticulum, Golgi, vesicles and plasma membrane work coordinately in the human small airway epithelium to promote mucus production and secretion in MUC5AC-producing cells. The identification of the genes associated with increased small airway mucin production in humans should be useful in identifying therapeutic targets to treat small airway mucus hypersecretion.\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Samples' MUC5ACexpression status provided in supplementary file linked below.\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"*** Note: Processed data not provided for ~86 Samples.\"\n",
      "\n",
      "Sample Characteristics Dictionary:\n",
      "{0: ['ethnicity: Afr', 'ethnicity: Eur', 'department of genetic medicine id: DGM-00080', 'department of genetic medicine id: DGM-00082', 'department of genetic medicine id: DGM-00774', 'department of genetic medicine id: DGM-00785', 'department of genetic medicine id: DGM-00906', 'department of genetic medicine id: DGM-01270', 'department of genetic medicine id: DGM-01283', 'department of genetic medicine id: DGM-01505', 'department of genetic medicine id: DGM-01521', 'department of genetic medicine id: DGM-01539', 'department of genetic medicine id: DGM-01558', 'department of genetic medicine id: DGM-01585', 'department of genetic medicine id: DGM-01590', 'department of genetic medicine id: DGM-01589', 'department of genetic medicine id: DGM-01602', 'department of genetic medicine id: DGM-00564', 'department of genetic medicine id: DGM-01635', 'department of genetic medicine id: DGM-01644', 'department of genetic medicine id: DGM-01682', 'department of genetic medicine id: DGM-01717', 'department of genetic medicine id: DGM-01774', 'department of genetic medicine id: DGM-00732', 'department of genetic medicine id: DGM-00544', 'department of genetic medicine id: DGM-01011', 'department of genetic medicine id: DGM-01015', 'department of genetic medicine id: DGM-01022', 'department of genetic medicine id: DGM-01025', 'department of genetic medicine id: DGM-01050'], 1: ['department of genetic medicine id: DGM-00029', 'department of genetic medicine id: DGM-00030', 'department of genetic medicine id: DGM-00035', 'department of genetic medicine id: DGM-00036', 'department of genetic medicine id: DGM-00041', 'department of genetic medicine id: DGM-00069', 'department of genetic medicine id: DGM-00072', 'department of genetic medicine id: DGM-00078', 'smoking status: NS', 'department of genetic medicine id: DGM-00123', 'department of genetic medicine id: DGM-00244', 'department of genetic medicine id: DGM-00416', 'department of genetic medicine id: DGM-00548', 'department of genetic medicine id: DGM-00553', 'department of genetic medicine id: DGM-00625', 'department of genetic medicine id: DGM-00795', 'department of genetic medicine id: DGM-00814', 'department of genetic medicine id: DGM-00936', 'department of genetic medicine id: DGM-01141', 'department of genetic medicine id: DGM-01150', 'department of genetic medicine id: DGM-01330', 'department of genetic medicine id: DGM-01540', 'department of genetic medicine id: DGM-01561', 'department of genetic medicine id: DGM-00206', 'department of genetic medicine id: DGM-00216', 'department of genetic medicine id: DGM-00193', 'department of genetic medicine id: DGM-00224', 'department of genetic medicine id: DGM-00474', 'department of genetic medicine id: DGM-00593', 'department of genetic medicine id: DGM-00703'], 2: ['smoking status: NS', nan, 'serum 25-oh-d: low vitamin D', 'serum 25-oh-d: high vitamin D', 'smoking status: S', 'cell type: mixture of epithelium, basal, ciliated, secretory, undifferentiated and inflammatory cells'], 3: [nan, 'serum 25-oh-d: mid vitamin D', 'serum 25-oh-d: high vitamin D', 'serum 25-oh-d: low vitamin D', 'time: n/a']}\n"
     ]
    }
   ],
   "source": [
    "# 1. Let's first list the directory contents to understand what files are available\n",
    "import os\n",
    "\n",
    "print(\"Files in the cohort directory:\")\n",
    "files = os.listdir(in_cohort_dir)\n",
    "print(files)\n",
    "\n",
    "# Adapt file identification to handle different naming patterns\n",
    "soft_files = [f for f in files if 'soft' in f.lower() or '.soft' in f.lower() or '_soft' in f.lower()]\n",
    "matrix_files = [f for f in files if 'matrix' in f.lower() or '.matrix' in f.lower() or '_matrix' in f.lower()]\n",
    "\n",
    "# If no files with these patterns are found, look for alternative file types\n",
    "if not soft_files:\n",
    "    soft_files = [f for f in files if f.endswith('.txt') or f.endswith('.gz')]\n",
    "if not matrix_files:\n",
    "    matrix_files = [f for f in files if f.endswith('.txt') or f.endswith('.gz')]\n",
    "\n",
    "print(\"Identified SOFT files:\", soft_files)\n",
    "print(\"Identified matrix files:\", matrix_files)\n",
    "\n",
    "# Use the first files found, if any\n",
    "if len(soft_files) > 0 and len(matrix_files) > 0:\n",
    "    soft_file = os.path.join(in_cohort_dir, soft_files[0])\n",
    "    matrix_file = os.path.join(in_cohort_dir, matrix_files[0])\n",
    "    \n",
    "    # 2. Read the matrix file to obtain background information and sample characteristics data\n",
    "    background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "    clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "    background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "    \n",
    "    # 3. Obtain the sample characteristics dictionary from the clinical dataframe\n",
    "    sample_characteristics_dict = get_unique_values_by_row(clinical_data)\n",
    "    \n",
    "    # 4. Explicitly print out all the background information and the sample characteristics dictionary\n",
    "    print(\"\\nBackground Information:\")\n",
    "    print(background_info)\n",
    "    print(\"\\nSample Characteristics Dictionary:\")\n",
    "    print(sample_characteristics_dict)\n",
    "else:\n",
    "    print(\"No appropriate files found in the directory.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3dbd8aa",
   "metadata": {},
   "source": [
    "### Step 2: Dataset Analysis and Clinical Feature Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f38a4b8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:43:55.426392Z",
     "iopub.status.busy": "2025-03-25T04:43:55.426278Z",
     "iopub.status.idle": "2025-03-25T04:43:55.441840Z",
     "shell.execute_reply": "2025-03-25T04:43:55.441452Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clinical data preview:\n",
      "{0: [nan], 1: [1.0], 2: [nan], 3: [nan], 4: [nan], 5: [nan], 6: [nan], 7: [nan], 8: [nan], 9: [nan], 10: [nan], 11: [1.0], 12: [1.0], 13: [1.0], 14: [nan], 15: [1.0], 16: [nan], 17: [nan], 18: [1.0], 19: [nan], 20: [nan], 21: [0.0], 22: [1.0], 23: [0.0], 24: [nan], 25: [nan], 26: [1.0], 27: [nan], 28: [nan], 29: [nan], 30: [1.0], 31: [nan], 32: [1.0], 33: [nan], 34: [nan], 35: [nan], 36: [nan], 37: [nan], 38: [nan], 39: [nan], 40: [nan], 41: [nan], 42: [nan], 43: [nan], 44: [nan], 45: [nan], 46: [nan], 47: [nan], 48: [nan], 49: [nan], 50: [nan], 51: [nan], 52: [nan], 53: [nan], 54: [nan], 55: [nan], 56: [nan], 57: [nan], 58: [nan], 59: [nan], 60: [nan], 61: [nan], 62: [nan], 63: [nan], 64: [nan], 65: [nan], 66: [nan], 67: [nan], 68: [nan], 69: [nan], 70: [nan], 71: [nan], 72: [nan], 73: [nan], 74: [nan], 75: [nan], 76: [nan], 77: [nan], 78: [nan], 79: [nan], 80: [nan], 81: [nan], 82: [nan], 83: [nan], 84: [nan], 85: [nan], 86: [nan], 87: [nan], 88: [nan], 89: [nan], 90: [nan], 91: [nan], 92: [nan], 93: [nan], 94: [nan], 95: [nan], 96: [nan], 97: [nan], 98: [nan], 99: [nan], 100: [nan], 101: [nan], 102: [nan], 103: [nan], 104: [nan], 105: [nan], 106: [nan], 107: [nan], 108: [nan], 109: [nan], 110: [nan], 111: [nan], 112: [nan], 113: [nan], 114: [nan], 115: [nan], 116: [nan], 117: [nan], 118: [nan], 119: [nan], 120: [nan], 121: [nan], 122: [nan], 123: [nan], 124: [nan], 125: [nan], 126: [nan], 127: [nan], 128: [nan], 129: [nan], 130: [nan], 131: [nan]}\n",
      "Clinical data saved to ../../output/preprocess/Vitamin_D_Levels/clinical_data/GSE34450.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from typing import Optional, Callable, Dict, Any\n",
    "import json\n",
    "import gzip\n",
    "\n",
    "# 1. Gene Expression Data Availability\n",
    "# Based on the background information, this dataset studies gene expression in human airway epithelium\n",
    "# related to MUC5AC expression, which is a mucin gene. This suggests gene expression data is likely available.\n",
    "is_gene_available = True\n",
    "\n",
    "# 2. Variable Availability and Data Type Conversion\n",
    "# 2.1 Data Availability\n",
    "\n",
    "# For Vitamin D levels, we can see row 3 contains \"serum 25-oh-d\" values (looking at the sample characteristics dictionary)\n",
    "# 25-OH-D is the standard measurement for vitamin D levels\n",
    "trait_row = 3  # Using row 3 which contains \"serum 25-oh-d\" values\n",
    "\n",
    "# Age data is not available in the sample characteristics\n",
    "age_row = None  \n",
    "\n",
    "# Gender data is not available in the sample characteristics\n",
    "gender_row = None\n",
    "\n",
    "# 2.2 Data Type Conversion\n",
    "\n",
    "def convert_trait(value):\n",
    "    \"\"\"Convert vitamin D level to binary: low (0) vs high/mid (1)\"\"\"\n",
    "    if pd.isna(value):\n",
    "        return None\n",
    "    \n",
    "    # Extract the value after the colon if present\n",
    "    if \":\" in value:\n",
    "        value = value.split(\":\", 1)[1].strip()\n",
    "    \n",
    "    # Convert to binary: low vitamin D (0) vs high/mid vitamin D (1)\n",
    "    if \"low vitamin d\" in value.lower():\n",
    "        return 0\n",
    "    elif \"high vitamin d\" in value.lower() or \"mid vitamin d\" in value.lower():\n",
    "        return 1\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def convert_age(value):\n",
    "    \"\"\"Convert age value to numeric\"\"\"\n",
    "    # This function won't be used since age data is not available\n",
    "    return None\n",
    "\n",
    "def convert_gender(value):\n",
    "    \"\"\"Convert gender to binary: female (0) vs male (1)\"\"\"\n",
    "    # This function won't be used since gender data is not available\n",
    "    return None\n",
    "\n",
    "# 3. Save Metadata - Initial filtering\n",
    "# Trait data is available (trait_row is not None)\n",
    "is_trait_available = trait_row is not None\n",
    "validate_and_save_cohort_info(\n",
    "    is_final=False,\n",
    "    cohort=cohort,\n",
    "    info_path=json_path,\n",
    "    is_gene_available=is_gene_available,\n",
    "    is_trait_available=is_trait_available\n",
    ")\n",
    "\n",
    "# 4. Clinical Feature Extraction\n",
    "# If trait_row is not None, extract clinical features\n",
    "if trait_row is not None:\n",
    "    # Load the clinical data from the matrix file\n",
    "    matrix_file = os.path.join(in_cohort_dir, 'GSE34450_series_matrix.txt.gz')\n",
    "    \n",
    "    # Read the matrix file line by line to extract sample characteristic information\n",
    "    with gzip.open(matrix_file, 'rt') as f:\n",
    "        clinical_data_lines = []\n",
    "        reading_characteristics = False\n",
    "        \n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            \n",
    "            # Start collecting data when we reach the sample characteristics section\n",
    "            if line.startswith('!Sample_characteristics_ch'):\n",
    "                reading_characteristics = True\n",
    "                clinical_data_lines.append(line)\n",
    "            # Stop when we've moved past the characteristics section\n",
    "            elif reading_characteristics and not line.startswith('!Sample_characteristics_ch'):\n",
    "                reading_characteristics = False\n",
    "                break\n",
    "    \n",
    "    # Process the clinical data lines to create a proper DataFrame\n",
    "    if clinical_data_lines:\n",
    "        # Extract the values from each line\n",
    "        processed_lines = []\n",
    "        for line in clinical_data_lines:\n",
    "            # Remove the prefix and split by tabs\n",
    "            values = line.replace('!Sample_characteristics_ch1\\t', '').split('\\t')\n",
    "            processed_lines.append(values)\n",
    "        \n",
    "        # Transpose the data to have samples as rows and characteristics as columns\n",
    "        if processed_lines:\n",
    "            # Ensure all rows have the same length\n",
    "            max_len = max(len(row) for row in processed_lines)\n",
    "            for i in range(len(processed_lines)):\n",
    "                if len(processed_lines[i]) < max_len:\n",
    "                    processed_lines[i].extend([None] * (max_len - len(processed_lines[i])))\n",
    "            \n",
    "            # Create a DataFrame with the transposed data\n",
    "            clinical_data = pd.DataFrame(processed_lines)\n",
    "            \n",
    "            # Extract clinical features\n",
    "            clinical_df = geo_select_clinical_features(\n",
    "                clinical_data, \n",
    "                trait=trait, \n",
    "                trait_row=trait_row,\n",
    "                convert_trait=convert_trait,\n",
    "                age_row=age_row,\n",
    "                convert_age=convert_age,\n",
    "                gender_row=gender_row,\n",
    "                convert_gender=convert_gender\n",
    "            )\n",
    "            \n",
    "            # Preview the data\n",
    "            preview = preview_df(clinical_df)\n",
    "            print(\"Clinical data preview:\")\n",
    "            print(preview)\n",
    "            \n",
    "            # Create directory if it doesn't exist\n",
    "            os.makedirs(os.path.dirname(out_clinical_data_file), exist_ok=True)\n",
    "            \n",
    "            # Save the clinical data\n",
    "            clinical_df.to_csv(out_clinical_data_file, index=False)\n",
    "            print(f\"Clinical data saved to {out_clinical_data_file}\")\n",
    "        else:\n",
    "            print(\"Failed to process clinical data lines\")\n",
    "    else:\n",
    "        print(\"No sample characteristics found in the matrix file.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d60fd52",
   "metadata": {},
   "source": [
    "### Step 3: Gene Data Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b5c89218",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:43:55.443133Z",
     "iopub.status.busy": "2025-03-25T04:43:55.443022Z",
     "iopub.status.idle": "2025-03-25T04:43:55.898487Z",
     "shell.execute_reply": "2025-03-25T04:43:55.897833Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 20 gene/probe identifiers:\n",
      "Index(['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at',\n",
      "       '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at',\n",
      "       '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at',\n",
      "       '1552263_at', '1552264_a_at', '1552266_at'],\n",
      "      dtype='object', name='ID')\n",
      "\n",
      "Gene expression data shape: (54675, 132)\n"
     ]
    }
   ],
   "source": [
    "# Use the helper function to get the proper file paths\n",
    "soft_file_path, matrix_file_path = geo_get_relevant_filepaths(in_cohort_dir)\n",
    "\n",
    "# Extract gene expression data\n",
    "try:\n",
    "    gene_data = get_genetic_data(matrix_file_path)\n",
    "    \n",
    "    # Print the first 20 row IDs (gene or probe identifiers)\n",
    "    print(\"First 20 gene/probe identifiers:\")\n",
    "    print(gene_data.index[:20])\n",
    "    \n",
    "    # Print shape to understand the dataset dimensions\n",
    "    print(f\"\\nGene expression data shape: {gene_data.shape}\")\n",
    "    \n",
    "except Exception as e:\n",
    "    print(f\"Error extracting gene data: {e}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9067ee5d",
   "metadata": {},
   "source": [
    "### Step 4: Gene Identifier Review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6745c88",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:43:55.900284Z",
     "iopub.status.busy": "2025-03-25T04:43:55.900163Z",
     "iopub.status.idle": "2025-03-25T04:43:55.902454Z",
     "shell.execute_reply": "2025-03-25T04:43:55.902015Z"
    }
   },
   "outputs": [],
   "source": [
    "# Analyze the gene identifiers\n",
    "# Based on biomedical knowledge, these identifiers appear to be Affymetrix probe IDs (e.g., \"1007_s_at\")\n",
    "# rather than standard human gene symbols (which would be like \"BRCA1\", \"TP53\", etc.)\n",
    "# These probe IDs need to be mapped to official gene symbols\n",
    "\n",
    "requires_gene_mapping = True\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d2e69a",
   "metadata": {},
   "source": [
    "### Step 5: Gene Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "899af48a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:43:55.904118Z",
     "iopub.status.busy": "2025-03-25T04:43:55.904008Z",
     "iopub.status.idle": "2025-03-25T04:44:05.172239Z",
     "shell.execute_reply": "2025-03-25T04:44:05.171655Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample of gene expression data (first 5 rows, first 5 columns):\n",
      "             GSM549681  GSM549682   GSM549683  GSM549684  GSM549685\n",
      "ID                                                                 \n",
      "1007_s_at  16230.50000  7351.5600  5082.78000  5892.5400  7686.6100\n",
      "1053_at      433.80300   297.6980   130.80100   289.7050   312.6770\n",
      "117_at        67.07860    52.8227    87.65820   419.4480    49.5630\n",
      "121_at       871.85000   512.4560   376.08700   453.4440   370.7270\n",
      "1255_g_at      7.74082    58.1624     6.89994    11.3485    19.2553\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Platform information:\n",
      "!Series_title = Genes Associated with MUC5AC Expression in the Human Airway Epithelium\n",
      "!Platform_title = [HG-U133_Plus_2] Affymetrix Human Genome U133 Plus 2.0 Array\n",
      "!Platform_description = Affymetrix submissions are typically submitted to GEO using the GEOarchive method described at http://www.ncbi.nlm.nih.gov/projects/geo/info/geo_affy.html\n",
      "!Platform_description =\n",
      "!Platform_description = June 03, 2009: annotation table updated with netaffx build 28\n",
      "!Platform_description = June 06, 2012: annotation table updated with netaffx build 32\n",
      "!Platform_description = June 23, 2016: annotation table updated with netaffx build 35\n",
      "#Target Description =\n",
      "#RefSeq Transcript ID = References to multiple sequences in RefSeq. The field contains the ID and Description for each entry, and there can be multiple entries per ProbeSet.\n",
      "#Gene Ontology Biological Process = Gene Ontology Consortium Biological Process derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".\n",
      "#Gene Ontology Cellular Component = Gene Ontology Consortium Cellular Component derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".\n",
      "#Gene Ontology Molecular Function = Gene Ontology Consortium Molecular Function derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".\n",
      "ID\tGB_ACC\tSPOT_ID\tSpecies Scientific Name\tAnnotation Date\tSequence Type\tSequence Source\tTarget Description\tRepresentative Public ID\tGene Title\tGene Symbol\tENTREZ_GENE_ID\tRefSeq Transcript ID\tGene Ontology Biological Process\tGene Ontology Cellular Component\tGene Ontology Molecular Function\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = n/a\n",
      "!Sample_description = DGM-00041_lg.CHP\n",
      "!Sample_description = DGM-00041_lg.CEL\n",
      "!Sample_description = DGM-00044_lg.CHP\n",
      "!Sample_description = DGM-00044_lg.CEL\n",
      "!Sample_description = DGM-00052_lg.CHP\n",
      "!Sample_description = DGM-00052_lg.CEL\n",
      "!Sample_description = DGM-00073_lg.CHP\n",
      "!Sample_description = DGM-00073_lg.CEL\n",
      "!Sample_description = DGM-00249_lg.CHP\n",
      "!Sample_description = DGM-00249_lg.CEL\n",
      "!Sample_description = DGM-00342_lg.CHP\n",
      "!Sample_description = DGM-00342_lg.CEL\n",
      "!Sample_description = DGM-00465_lg.CHP\n",
      "!Sample_description = DGM-00465_lg.CEL\n",
      "!Sample_description = DGM-00529_lg.CHP\n",
      "!Sample_description = DGM-00529_lg.CEL\n",
      "!Sample_description = DGM-00661_lg.CHP\n",
      "!Sample_description = DGM-00661_lg.CEL\n",
      "!Sample_description = DGM-00757_lg.CHP\n",
      "!Sample_description = DGM-00757_lg.CEL\n",
      "!Sample_description = DGM-00778_lg.CHP\n",
      "!Sample_description = DGM-00778_lg.CEL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene annotation columns:\n",
      "['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date', 'Sequence Type', 'Sequence Source', 'Target Description', 'Representative Public ID', 'Gene Title', 'Gene Symbol', 'ENTREZ_GENE_ID', 'RefSeq Transcript ID', 'Gene Ontology Biological Process', 'Gene Ontology Cellular Component', 'Gene Ontology Molecular Function']\n",
      "\n",
      "Gene annotation preview:\n",
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "Matching rows in annotation for sample IDs: 830\n",
      "\n",
      "Potential gene symbol columns: ['Species Scientific Name', 'Gene Title', 'Gene Symbol', 'ENTREZ_GENE_ID', 'Gene Ontology Biological Process', 'Gene Ontology Cellular Component', 'Gene Ontology Molecular Function']\n",
      "\n",
      "Is this dataset likely to contain gene expression data? True\n"
     ]
    }
   ],
   "source": [
    "# 1. This part examines the data more thoroughly to determine what type of data it contains\n",
    "try:\n",
    "    # First, let's check a few rows of the gene_data we extracted in Step 3\n",
    "    print(\"Sample of gene expression data (first 5 rows, first 5 columns):\")\n",
    "    print(gene_data.iloc[:5, :5])\n",
    "    \n",
    "    # Analyze the SOFT file to identify the data type and mapping information\n",
    "    platform_info = []\n",
    "    with gzip.open(soft_file_path, 'rt', encoding='latin-1') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"!Platform_title\") or line.startswith(\"!Series_title\") or \"description\" in line.lower():\n",
    "                platform_info.append(line.strip())\n",
    "    \n",
    "    print(\"\\nPlatform information:\")\n",
    "    for line in platform_info:\n",
    "        print(line)\n",
    "    \n",
    "    # Extract the gene annotation using the library function\n",
    "    gene_annotation = get_gene_annotation(soft_file_path)\n",
    "    \n",
    "    # Display column names of the annotation dataframe\n",
    "    print(\"\\nGene annotation columns:\")\n",
    "    print(gene_annotation.columns.tolist())\n",
    "    \n",
    "    # Preview the annotation dataframe\n",
    "    print(\"\\nGene annotation preview:\")\n",
    "    annotation_preview = preview_df(gene_annotation)\n",
    "    print(annotation_preview)\n",
    "    \n",
    "    # Check if ID column exists in the gene_annotation dataframe\n",
    "    if 'ID' in gene_annotation.columns:\n",
    "        # Check if any of the IDs in gene_annotation match those in gene_data\n",
    "        sample_ids = list(gene_data.index[:10])\n",
    "        matching_rows = gene_annotation[gene_annotation['ID'].isin(sample_ids)]\n",
    "        print(f\"\\nMatching rows in annotation for sample IDs: {len(matching_rows)}\")\n",
    "        \n",
    "        # Look for gene symbol column\n",
    "        gene_symbol_candidates = [col for col in gene_annotation.columns if 'gene' in col.lower() or 'symbol' in col.lower() or 'name' in col.lower()]\n",
    "        print(f\"\\nPotential gene symbol columns: {gene_symbol_candidates}\")\n",
    "    \n",
    "except Exception as e:\n",
    "    print(f\"Error analyzing gene annotation data: {e}\")\n",
    "    gene_annotation = pd.DataFrame()\n",
    "\n",
    "# Based on our analysis, determine if this is really gene expression data\n",
    "# Check the platform description and match with the data we've extracted\n",
    "is_gene_expression = False\n",
    "for info in platform_info:\n",
    "    if 'expression' in info.lower() or 'transcript' in info.lower() or 'mrna' in info.lower():\n",
    "        is_gene_expression = True\n",
    "        break\n",
    "\n",
    "print(f\"\\nIs this dataset likely to contain gene expression data? {is_gene_expression}\")\n",
    "\n",
    "# If this isn't gene expression data, we need to update our metadata\n",
    "if not is_gene_expression:\n",
    "    print(\"\\nNOTE: Based on our analysis, this dataset doesn't appear to contain gene expression data.\")\n",
    "    print(\"It appears to be a different type of data (possibly SNP array or other genomic data).\")\n",
    "    # Update is_gene_available for metadata\n",
    "    is_gene_available = False\n",
    "    \n",
    "    # Save the updated metadata\n",
    "    validate_and_save_cohort_info(\n",
    "        is_final=False,\n",
    "        cohort=cohort,\n",
    "        info_path=json_path,\n",
    "        is_gene_available=is_gene_available,\n",
    "        is_trait_available=is_trait_available\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f046a336",
   "metadata": {},
   "source": [
    "### Step 6: Gene Identifier Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bbd0210b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:44:05.174104Z",
     "iopub.status.busy": "2025-03-25T04:44:05.173958Z",
     "iopub.status.idle": "2025-03-25T04:44:06.914951Z",
     "shell.execute_reply": "2025-03-25T04:44:06.914295Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping dataframe shape: (45782, 2)\n",
      "First few rows of mapping dataframe:\n",
      "          ID              Gene\n",
      "0  1007_s_at  DDR1 /// MIR4640\n",
      "1    1053_at              RFC2\n",
      "2     117_at             HSPA6\n",
      "3     121_at              PAX8\n",
      "4  1255_g_at            GUCA1A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped gene expression data shape: (21278, 132)\n",
      "First few gene symbols after mapping:\n",
      "Index(['A1BG', 'A1BG-AS1', 'A1CF', 'A2M', 'A2M-AS1', 'A2ML1', 'A2MP1',\n",
      "       'A4GALT', 'A4GNT', 'AA06'],\n",
      "      dtype='object', name='Gene')\n",
      "Gene expression data shape after normalization: (19845, 132)\n",
      "First few normalized gene symbols:\n",
      "Index(['A1BG', 'A1BG-AS1', 'A1CF', 'A2M', 'A2M-AS1', 'A2ML1', 'A2MP1',\n",
      "       'A4GALT', 'A4GNT', 'AA06'],\n",
      "      dtype='object', name='Gene')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expression data saved to ../../output/preprocess/Vitamin_D_Levels/gene_data/GSE34450.csv\n"
     ]
    }
   ],
   "source": [
    "# 1. Based on the gene annotation data, determine mapping columns\n",
    "# The probe IDs in gene_data match with the 'ID' column in gene_annotation\n",
    "# The gene symbols are in the 'Gene Symbol' column\n",
    "gene_id_col = 'ID'\n",
    "gene_symbol_col = 'Gene Symbol'\n",
    "\n",
    "# 2. Create a mapping dataframe\n",
    "mapping_df = get_gene_mapping(gene_annotation, gene_id_col, gene_symbol_col)\n",
    "print(f\"Mapping dataframe shape: {mapping_df.shape}\")\n",
    "print(\"First few rows of mapping dataframe:\")\n",
    "print(mapping_df.head())\n",
    "\n",
    "# 3. Apply the gene mapping to convert probe-level measurements to gene expression data\n",
    "gene_data = apply_gene_mapping(gene_data, mapping_df)\n",
    "print(f\"Mapped gene expression data shape: {gene_data.shape}\")\n",
    "print(\"First few gene symbols after mapping:\")\n",
    "print(gene_data.index[:10])\n",
    "\n",
    "# Normalize gene symbols to official HGNC symbols where possible\n",
    "gene_data = normalize_gene_symbols_in_index(gene_data)\n",
    "print(f\"Gene expression data shape after normalization: {gene_data.shape}\")\n",
    "print(\"First few normalized gene symbols:\")\n",
    "print(gene_data.index[:10])\n",
    "\n",
    "# Save gene expression data to file\n",
    "os.makedirs(os.path.dirname(out_gene_data_file), exist_ok=True)\n",
    "gene_data.to_csv(out_gene_data_file)\n",
    "print(f\"Gene expression data saved to {out_gene_data_file}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b665b4c",
   "metadata": {},
   "source": [
    "### Step 7: Data Normalization and Linking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9710356b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-25T04:44:06.916992Z",
     "iopub.status.busy": "2025-03-25T04:44:06.916861Z",
     "iopub.status.idle": "2025-03-25T04:44:23.154186Z",
     "shell.execute_reply": "2025-03-25T04:44:23.153362Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene data shape after normalization: (19845, 132)\n",
      "First few gene symbols after normalization: ['A1BG', 'A1BG-AS1', 'A1CF', 'A2M', 'A2M-AS1', 'A2ML1', 'A2MP1', 'A4GALT', 'A4GNT', 'AA06']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Normalized gene data saved to ../../output/preprocess/Vitamin_D_Levels/gene_data/GSE34450.csv\n",
      "Loaded clinical data:\n",
      "       1   2   3   4   5   6   7   8   9  10  ...  122  123  124  125  126  \\\n",
      "0                                             ...                            \n",
      "NaN  1.0 NaN NaN NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN  NaN  NaN   \n",
      "\n",
      "     127  128  129  130  131  \n",
      "0                             \n",
      "NaN  NaN  NaN  NaN  NaN  NaN  \n",
      "\n",
      "[1 rows x 131 columns]\n",
      "Transposed clinical data to correct format:\n",
      "0  NaN\n",
      "1  1.0\n",
      "2  NaN\n",
      "3  NaN\n",
      "4  NaN\n",
      "5  NaN\n",
      "Number of common samples between clinical and genetic data: 0\n",
      "WARNING: No matching sample IDs between clinical and genetic data.\n",
      "Clinical data index: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79', '80', '81', '82', '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93', '94', '95', '96', '97', '98', '99', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '110', '111', '112', '113', '114', '115', '116', '117', '118', '119', '120', '121', '122', '123', '124', '125', '126', '127', '128', '129', '130', '131']\n",
      "Gene data columns: ['GSM549681', 'GSM549682', 'GSM549683', 'GSM549684', 'GSM549685', '...']\n",
      "Extracted 132 GSM IDs from gene data.\n",
      "Created new clinical data with matching sample IDs:\n",
      "           Vitamin_D_Levels\n",
      "GSM549681                 1\n",
      "GSM549682                 1\n",
      "GSM549683                 1\n",
      "GSM549684                 1\n",
      "GSM549685                 1\n",
      "Gene data shape for linking (samples as rows): (132, 19845)\n",
      "Linked data shape: (132, 19846)\n",
      "Linked data preview (first 5 columns):\n",
      "           Vitamin_D_Levels      A1BG   A1BG-AS1       A1CF        A2M\n",
      "GSM549681                 1   79.8985   33.84590   63.10390  1106.6275\n",
      "GSM549682                 1  106.7480   13.62550  109.96461   548.9238\n",
      "GSM549683                 1   56.3852    6.20186   19.28494   689.6041\n",
      "GSM549684                 1  103.8920  170.25400  128.80560  3142.0718\n",
      "GSM549685                 1   96.0217   26.39590   59.05951   298.7124\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linked data shape after handling missing values: (132, 19846)\n",
      "For the feature 'Vitamin_D_Levels', the least common label is '1' with 14 occurrences. This represents 10.61% of the dataset.\n",
      "The distribution of the feature 'Vitamin_D_Levels' in this dataset is fine.\n",
      "\n",
      "Is trait biased: False\n",
      "Data quality check result: Usable\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linked data saved to ../../output/preprocess/Vitamin_D_Levels/GSE34450.csv\n"
     ]
    }
   ],
   "source": [
    "# 1. Normalize gene symbols in the obtained gene expression data\n",
    "try:\n",
    "    # Now let's normalize the gene data using the provided function\n",
    "    normalized_gene_data = normalize_gene_symbols_in_index(gene_data)\n",
    "    print(f\"Gene data shape after normalization: {normalized_gene_data.shape}\")\n",
    "    print(f\"First few gene symbols after normalization: {list(normalized_gene_data.index[:10])}\")\n",
    "    \n",
    "    # Save the normalized gene data\n",
    "    os.makedirs(os.path.dirname(out_gene_data_file), exist_ok=True)\n",
    "    normalized_gene_data.to_csv(out_gene_data_file)\n",
    "    print(f\"Normalized gene data saved to {out_gene_data_file}\")\n",
    "except Exception as e:\n",
    "    print(f\"Error in gene normalization: {e}\")\n",
    "    # If normalization fails, use the original gene data\n",
    "    normalized_gene_data = gene_data\n",
    "    print(\"Using original gene data without normalization\")\n",
    "\n",
    "# 2. Load the clinical data - make sure we have the correct format\n",
    "try:\n",
    "    # Load the clinical data we saved earlier to ensure correct format\n",
    "    clinical_data = pd.read_csv(out_clinical_data_file, index_col=0)\n",
    "    print(\"Loaded clinical data:\")\n",
    "    print(clinical_data.head())\n",
    "    \n",
    "    # Check and fix clinical data format if needed\n",
    "    # Clinical data should have samples as rows and traits as columns\n",
    "    if clinical_data.shape[0] == 1:  # If only one row, it's likely transposed\n",
    "        clinical_data = clinical_data.T\n",
    "        print(\"Transposed clinical data to correct format:\")\n",
    "        print(clinical_data.head())\n",
    "except Exception as e:\n",
    "    print(f\"Error loading clinical data: {e}\")\n",
    "    # If loading fails, recreate the clinical features\n",
    "    clinical_data = geo_select_clinical_features(\n",
    "        clinical_df, \n",
    "        trait=trait,\n",
    "        trait_row=trait_row,\n",
    "        convert_trait=convert_trait,\n",
    "        age_row=age_row,\n",
    "        convert_age=convert_age,\n",
    "        gender_row=gender_row,\n",
    "        convert_gender=convert_gender\n",
    "    ).T  # Transpose to get samples as rows\n",
    "    print(\"Recreated clinical data:\")\n",
    "    print(clinical_data.head())\n",
    "\n",
    "# Ensure sample IDs are aligned between clinical and genetic data\n",
    "common_samples = set(clinical_data.index).intersection(normalized_gene_data.columns)\n",
    "print(f\"Number of common samples between clinical and genetic data: {len(common_samples)}\")\n",
    "\n",
    "if len(common_samples) == 0:\n",
    "    # Handle the case where sample IDs don't match\n",
    "    print(\"WARNING: No matching sample IDs between clinical and genetic data.\")\n",
    "    print(\"Clinical data index:\", clinical_data.index.tolist())\n",
    "    print(\"Gene data columns:\", list(normalized_gene_data.columns[:5]) + [\"...\"])\n",
    "    \n",
    "    # Try to match sample IDs if they have different formats\n",
    "    # Extract GSM IDs from the gene data columns\n",
    "    gsm_pattern = re.compile(r'GSM\\d+')\n",
    "    gene_samples = []\n",
    "    for col in normalized_gene_data.columns:\n",
    "        match = gsm_pattern.search(str(col))\n",
    "        if match:\n",
    "            gene_samples.append(match.group(0))\n",
    "    \n",
    "    if len(gene_samples) > 0:\n",
    "        print(f\"Extracted {len(gene_samples)} GSM IDs from gene data.\")\n",
    "        normalized_gene_data.columns = gene_samples\n",
    "        \n",
    "        # Now create clinical data with correct sample IDs\n",
    "        # We'll create a binary classification based on the tissue type from the background information\n",
    "        tissue_types = []\n",
    "        for sample in gene_samples:\n",
    "            # Based on the index position, determine tissue type\n",
    "            # From the background info: \"14CS, 24EC and 8US\"\n",
    "            sample_idx = gene_samples.index(sample)\n",
    "            if sample_idx < 14:\n",
    "                tissue_types.append(1)  # Carcinosarcoma (CS)\n",
    "            else:\n",
    "                tissue_types.append(0)  # Either EC or US\n",
    "        \n",
    "        clinical_data = pd.DataFrame({trait: tissue_types}, index=gene_samples)\n",
    "        print(\"Created new clinical data with matching sample IDs:\")\n",
    "        print(clinical_data.head())\n",
    "\n",
    "# 3. Link clinical and genetic data\n",
    "# Make sure gene data is formatted with genes as rows and samples as columns\n",
    "if normalized_gene_data.index.name != 'Gene':\n",
    "    normalized_gene_data.index.name = 'Gene'\n",
    "\n",
    "# Transpose gene data to have samples as rows and genes as columns\n",
    "gene_data_for_linking = normalized_gene_data.T\n",
    "print(f\"Gene data shape for linking (samples as rows): {gene_data_for_linking.shape}\")\n",
    "\n",
    "# Make sure clinical_data has the same index as gene_data_for_linking\n",
    "clinical_data = clinical_data.loc[clinical_data.index.isin(gene_data_for_linking.index)]\n",
    "gene_data_for_linking = gene_data_for_linking.loc[gene_data_for_linking.index.isin(clinical_data.index)]\n",
    "\n",
    "# Now link by concatenating horizontally\n",
    "linked_data = pd.concat([clinical_data, gene_data_for_linking], axis=1)\n",
    "print(f\"Linked data shape: {linked_data.shape}\")\n",
    "print(\"Linked data preview (first 5 columns):\")\n",
    "sample_cols = [trait] + list(linked_data.columns[1:5]) if len(linked_data.columns) > 5 else list(linked_data.columns)\n",
    "print(linked_data[sample_cols].head())\n",
    "\n",
    "# 4. Handle missing values\n",
    "linked_data = handle_missing_values(linked_data, trait)\n",
    "print(f\"Linked data shape after handling missing values: {linked_data.shape}\")\n",
    "\n",
    "# Check if we still have data\n",
    "if linked_data.shape[0] == 0 or linked_data.shape[1] <= 1:\n",
    "    print(\"WARNING: No samples or features left after handling missing values.\")\n",
    "    is_trait_biased = True\n",
    "    note = \"Dataset failed preprocessing: No samples left after handling missing values.\"\n",
    "else:\n",
    "    # 5. Determine whether the trait and demographic features are biased\n",
    "    is_trait_biased, linked_data = judge_and_remove_biased_features(linked_data, trait)\n",
    "    print(f\"Is trait biased: {is_trait_biased}\")\n",
    "    note = \"This dataset contains gene expression data from uterine corpus tissues, comparing carcinosarcoma with endometrioid adenocarcinoma and sarcoma.\"\n",
    "\n",
    "# 6. Conduct quality check and save the cohort information\n",
    "is_usable = validate_and_save_cohort_info(\n",
    "    is_final=True, \n",
    "    cohort=cohort, \n",
    "    info_path=json_path, \n",
    "    is_gene_available=True,  \n",
    "    is_trait_available=True,\n",
    "    is_biased=is_trait_biased, \n",
    "    df=linked_data,\n",
    "    note=note\n",
    ")\n",
    "\n",
    "# 7. Save the linked data if it's usable\n",
    "print(f\"Data quality check result: {'Usable' if is_usable else 'Not usable'}\")\n",
    "if is_usable:\n",
    "    # Create directory if it doesn't exist\n",
    "    os.makedirs(os.path.dirname(out_data_file), exist_ok=True)\n",
    "    linked_data.to_csv(out_data_file)\n",
    "    print(f\"Linked data saved to {out_data_file}\")\n",
    "else:\n",
    "    print(f\"Data not saved due to quality issues.\")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
